include/geode/basic/database.h

```diff
@@ -0,0 +1,112 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#pragma once
+
+#include <geode/basic/common.h>
+#include <geode/basic/logger.h>
+#include <geode/basic/pimpl.h>
+
+#include <geode/basic/identifier.h>
+#include <geode/basic/uuid.h>
+
+namespace geode
+{
+    class Identifier;
+    struct uuid;
+} // namespace geode
+
+namespace geode
+{
+    class opengeode_basic_api Database
+    {
+        OPENGEODE_DISABLE_COPY( Database );
+        struct Storage;
+
+    public:
+        using serializer_function = std::function< void( PContext& ) >;
+
+        class opengeode_basic_api Data
+        {
+        public:
+            Data( std::shared_ptr< Storage > storage );
+            ~Data();
+            Data( Data&& other );
+
+            template < typename DataType >
+            const DataType& get()
+            {
+                const auto& typed_data =
+                    dynamic_cast< const DataType& >( data() );
+                return typed_data;
+            }
+
+        private:
+            const Identifier& data() const;
+
+        private:
+            IMPLEMENTATION_MEMBER( impl_ );
+        };
+
+    public:
+        Database( absl::string_view directory );
+        ~Database();
+
+        index_t nb_data() const;
+
+        template < typename DataType >
+        const uuid& register_data( DataType&& data )
+        {
+            static_assert( std::is_base_of< Identifier, DataType >::value,
+                "[Database::register_data] Data is not a subclass of "
+                "Identifier" );
+            return register_unique_data(
+                absl::make_unique< DataType >( std::move( data ) ) );
+        }
+
+        template < typename DataType >
+        const uuid& register_data( std::unique_ptr< DataType >&& data )
+        {
+            static_assert( std::is_base_of< Identifier, DataType >::value,
+                "[Database::register_data] Data is not a subclass of "
+                "Identifier" );
+            return register_unique_data( std::move( data ) );
+        }
+
+        void delete_data( const uuid& id );
+
+        Data get_data( const uuid& id ) const;
+
+        std::unique_ptr< Identifier > take_data( const uuid& id );
```

Comment:
We could template this method BUT we need to make sure to handle cast failure properly. What behavior if a cast failed? Does the data is still taken and need a reload?

Reply:
empty or exception? I would prefer exception I think